Deploy dex in metalk8s from helm charts

Ebaneck opened the pull request:
**Component**:

'authentication', 'salt', 'kubernetes'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

See #2007 

**Summary**:




**Acceptance criteria**: 

- Have Dex pods running in MetalK8s cluster deployed from helm charts
- Access the dex URL exposed by ingress and verify that it works : https://{{ control_plane_address }}:8443/oidc/.well-known/openid-configuration

```
{
  "issuer": "https://172.21.254.3:8443/oidc",
  "authorization_endpoint": "https://172.21.254.3:8443/oidc/auth",
  "token_endpoint": "https://172.21.254.3:8443/oidc/token",
  "jwks_uri": "https://172.21.254.3:8443/oidc/keys",
  "userinfo_endpoint": "https://172.21.254.3:8443/oidc/userinfo",
  "response_types_supported": [
    "code",
    "id_token",
    "token"
  ],
  "subject_types_supported": [
    "public"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "scopes_supported": [
    "openid",
    "email",
    "groups",
    "profile",
    "offline_access"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic"
  ],
  "claims_supported": [
    "aud",
    "email",
    "email_verified",
    "exp",
    "iat",
    "iss",
    "locale",
    "name",
    "sub"
  ]
}
```

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2007 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+image: '{%- endraw -%}{{ build_image_name(\"dex\", False) }}{%- raw -%}'
+
+nodeSelector:
+  node-role.kubernetes.io/master: ''
```
No, Dex is infra, not control-plane.

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+  - key: "node-role.kubernetes.io/bootstrap"
+    operator: "Exists"
+    effect: "NoSchedule"
+  - key: "node-role.kubernetes.io/master"
```
I don't think we want a control-plane toleration, only `bootstrap` and `infra`.

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+grpc: false
+
+# https termination by dex itself
+https: false
```
Why not?

gdemonet replied:
Indeed, even if we expose Dex through an Ingress object, it would make sense to still have it secured and exposed using a `NodePort` (though not entirely sure about that part, cc @NicolasT), as we do for K8s and Salt APIs.

NicolasT replied:
No no no, we don't want/need a `NodePort`.

However, when we *can* secure in-cluster over-the-wire traffic, we should.

gdemonet replied:
OK OK OK 😄 No `NodePort` then!

Ebaneck replied:
Yeah,  I am going to enable TLS here

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+
+ports:
+  web:
+    containerPort: 5556
```
Do we need to override any defaults?

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+    # servicePort: 35000
+
+service:
+  type: ClusterIP
```
I'd assume this is the default as well, though didn't check.

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+    nginx.ingress.kubernetes.io/use-regex: "true"
+    nginx.ingress.kubernetes.io/backend-protocol: "HTTP"
+    kubernetes.io/ingress.class: "nginx-control-plane"
+  path: /dex(/|$)(.*)
```
I'd refrain from using `dex` in the URL (cfr. namespace thing). Consider `/oidc` or such?

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+  path: /dex(/|$)(.*)
+  hosts:
+    - null
+  tls: []
```
I don't think not setting one will work. The ingress will get a new cert at every start (which is another issue, but anyway), so the cert registered with clients will never be correct, isn't it?

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+    create: false
+
+rbac:
+  create: true
```
Isn't this the default as well? (Also applies to various other settings in this file)

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: "node-role.kubernetes.io/master"
```
Infra. And because we already have a `nodeSelector`, there should be no more affinity defined?

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+
+  enablePasswordDB: true
+  staticPasswords:
+    - email: "oidc_admin@metalk8s.com"
```
There is no such domain. Need to find a better value for this, similar to our fake registry address.

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,102 @@
+    - email: "oidc_admin@metalk8s.com"
+      # bcrypt hash of the string "password"
+      hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
+      username: "oidc_admin"
```
Why this username?

NicolasT commented on salt/metalk8s/addons/dex/deployed/namespace.sls:
```diff
@@ -0,0 +1,10 @@
+metadata:
+  name: metalk8s-auth
+  labels:
+    app.kubernetes.io/managed-by: metalk8s
```
`managed-by` `salt`

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

Ebaneck commented on charts/dex/templates/deployment.yaml:
```diff
@@ -0,0 +1,114 @@
+{{ toYaml . | indent 8 }}
+    {{- end }}
+      containers:
+      - name: main
```
@gdemonet this seems to be an issue from the Dex stable helm charts.
Dex Container names are hard-coded to `main`
I think we are not allowed to make changes to this file in the way we deploy using charts.. no ??

NicolasT replied:
Why would this name be an issue in the first place?

Ebaneck replied:
Not a big deal, maybe while listing the running containers, it looked a little inconsistent

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+  connectors: {}
+
+  staticPasswords:
+    - email: "admin@example.com"
```
I commented on this before but the comment seems to be gone?

Ebaneck replied:
It was not marked resolved yet

NicolasT replied:
Sure, but you did make a change to the line, and as such GitHub removed the comment from the diff view. Which can be considered an issue in the GitHub UI.

gdemonet replied:
What's this e-mail used for anyway? Could it be `admin@metalk8s.invalid` or something along those lines?

NicolasT replied:
The e-mail address is used in the OIDC profile of this 'user', and as such can show up e.g. in Grafana.

It should indeed be set to some 'invalid' address.

NicolasT commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+    create: false
+
+config:
+  issuer: https://{{ pillar['metalk8s']['api_server']['host'] }}:32000
```
I'm surprised this works without an `endraw`/`raw`.

Ebaneck replied:
Yeah it doen't

NicolasT commented on salt/metalk8s/addons/dex/deployed/chart.sls:
```diff
@@ -0,0 +1,139 @@
+    heritage: metalk8s}
+  name: dex
+  namespace: metalk8s-auth
+stringData: {config.yaml: "issuer: https://{{ pillar['metalk8s']['api_server']['host']\
```
As hinted above (in the values file), this ain't right. Even adding `endraw`/`raw` there may make this fail... Not sure how to fix though.

gdemonet replied:
`endraw`/`raw` should work, don't see why it wouldn't... Jinja doesn't care about what the supposed format underneath is, IIRC.

NicolasT replied:
It could not work because of the `\`s that get injected when serializing the YAML.

gdemonet replied:
Maybe with block-style display this wouldn't need to inject `\`s... We need to test, ping @Ebaneck ;)

Ebaneck replied:
I can confirm that endraw / raw does not work

Ebaneck replied:
Even with block-style display, it injects `\ s..` Any other hints i could try out?? 

Worst case, I add an sls that creates the same secret holding the dex config params

NicolasT commented on salt/metalk8s/addons/dex/deployed/chart.sls:
```diff
@@ -0,0 +1,139 @@
+      - backend: {serviceName: dex, servicePort: 32000}
+        path: /oidc(/|$)(.*)
+  tls:
+  - {hosts: null, secretName: dex-tls-secret}
```
Where is this created?

Ebaneck replied:
Working on this, I am going to add a `secret` that would hold `cert` and `key`

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+
+  web:
+    # port is taken from ports section above
+    address: 0.0.0.0
```
Shouldn't it listen on the control-plane IP only, as suggested in #2011?

NicolasT replied:
Huh, no.

This stuff is running in a container, not in the host network, so binding to a control-plane IP (which IP would that be, by the way? Of which node?) can't work.

`0.0.0.0` is fine and will basically bind to `lo` and the only other interface in the container.

However, again, isn't this the default in the chart's `values`, so do we really need to set it here?!

Ebaneck replied:
it's default in the charts.. taking it off

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+    operator: "Exists"
+    effect: "NoSchedule"
+
+replicas: 2
```
Why? The default is 1, and the initial deployment is on Bootstrap only. Would make sense to handle this as we do for Prom/AM, i.e. use 1 replica for now, and make it configurable somehow (that'd be part of the #1996 epic I guess).

NicolasT replied:
Dex is stateless, so having 2 shouldn't harm. Even better, given the `updateStrategy`, there will always be one running even during upgrades and such.

Initially both may run on the `bootstrap` node, fine, but once there are more `infra` ones things can be spread.

So I agree with 2 :)

gdemonet replied:
OK, I get why having more than one can help for `updateStrategy`. Would still be useful to make it configurable in the long run, IMHO.

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+replicas: 2
+
+# grpc support
+grpc: false
```
Didn't we say we'll need gRPC for #1990?

NicolasT replied:
Do we really need GRPC? I'm afraid of getting into an authn/authz  mess, whilst access to the configuration/configmap is much easier to manage.

gdemonet replied:
If it's acceptable to restart Dex for each config change, then yes avoid the whole gRPC stuff can be much simpler.

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,56 @@
+    kubernetes.io/ingress.class: "nginx-control-plane"
+  path: /oidc(/|$)(.*)
+  hosts:
+    - null
```
Control-plane IP here?

Ebaneck replied:
Not sure we can have an IP address in ingress's host

NicolasT replied:
No, not control plane IP... I'm not even sure `null` is right, maybe the lists needs to be empty?

In any case, we want to be the 'fallback' entry (like we do for UI, Grafana,...) for the ingress controller.

gdemonet replied:
Unfortunately, this may be hard to do using the Helm chart, given this [template](https://github.com/helm/charts/blob/master/stable/dex/templates/ingress.yaml#L27-L35):

```yaml
  rules:
  {{- range .Values.ingress.hosts }}
    - host: {{ . | quote }}
      http:
        paths:
          - path: {{ $ingressPath }}
            backend:
              serviceName: {{ $fullName }}
              servicePort: {{ $servicePort }}
  {{- end }}
```

gdemonet commented on salt/metalk8s/addons/dex/deployed/chart.sls:
```diff
@@ -0,0 +1,139 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  labels: {app.kubernetes.io/instance: dex, app.kubernetes.io/managed-by: salt, app.kubernetes.io/name: dex,
```
This display of YAML is unreadable for me, I'd rather use block style consistently. See: https://github.com/scality/metalk8s/pull/2024/commits/3cdd70704250346fc938d778311114414171543b

gdemonet commented on salt/metalk8s/addons/dex/deployed/dex-conf.sls:
```diff
@@ -0,0 +1,75 @@
+  namespace: metalk8s-auth
+stringData:
+  config.yaml: |-
+      issuer: 'https://{{ oidc_service_ip }}:32000'
```
What would it imply to mention use the Ingress in this config file? Does it only affect what is exposed through `.well-know/openid-configuration`?

```suggestion
      issuer: 'https://{{ grains.metalk8s.control_plane_ip }}:8443/oidc'
```

TeddyAndrieux commented on pillar/metalk8s/roles/ca.sls:
```diff
@@ -51,3 +59,10 @@ x509_signing_policies:
+    - signing_private_key: /etc/kubernetes/pki/dex-ca.key
+    - signing_cert: /etc/kubernetes/pki/dex-ca.crt
+    - keyUsage: critical digitalSignature, keyEncipherment
+    - extendedKeyUsage: serverAuth, clientAuth
```
Do we really need `clientAuth` Dex cert is only used as server no ?

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/2007-deploy-dex-in-metalk8s-from-helm-charts` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/2007-deploy-dex-in-metalk8s-from-helm-charts`
* Rebase `feature/2007-deploy-dex-in-metalk8s-from-helm-charts` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

gdemonet committed:
Charts: render multiline strings using block-style

This changed is added because multi-line strings is an issue in
rendered charts since Jinja "commands" are split.

This fix is needed to properly add and render Dex charts as done
in the next commit

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

TeddyAndrieux commented on charts/dex/templates/clusterrole.yaml:
```diff
@@ -9,7 +9,44 @@ rules:
+    - list
+    - create
+    - patch
+    - update
```
Wait what ??? Do we really want to have (almost) everything ??
If I well understand it's only stuff doable by the UI no ? So only listing/getting some stuff + maybe some edition no ?

gdemonet replied:
Agree with Teddy, this doesn't make sense. This ClusterRole is for Dex, so that it can use K8s API as its storage, it should have write permissions on such generic resources. Where does this come from? The `clusterrole` from upstream looks better: https://github.com/helm/charts/blob/master/stable/dex/templates/clusterrole.yaml

Ebaneck replied:
yeah, i am going to keep the upstream one

TeddyAndrieux commented on charts/dex.yaml:
```diff
@@ -0,0 +1,96 @@
+  annotations:
+    nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
+    nginx.ingress.kubernetes.io/proxy-ssl-verify: "on"
+    nginx.ingress.kubernetes.io/proxy-ssl-secret: "dex-ca-cert"
```
I think since we have a default cert for ingress we no longer need this dex CA and server cert

TeddyAndrieux commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -4,9 +4,40 @@
+{%- else %}
+  {%- set bootstrap_control_plane_ip = salt['mine.get'](bootstrap_id,
+                                                        'control_plane_ip') %}
+{%- endif %}
```
Why we want to use bootstrap ip, that mean if we lose the bootstrap node oidc will no longer work,

Also meaning that we need to update each and every apiserver in restore script

gdemonet replied:
We need a single IP, so a LoadBalancer IP would be ideal, but we don't have a way to create one yet (see #1788).
In the meantime, we use the only available externalIP exposed by the nginx-ingress-control-plane Service, which the Bootstrap's control-plane IP.

Indeed, it does mean that a lot of things will stop working if Bootstrap goes down. The solution is just not available at the moment.

gdemonet commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -4,9 +4,40 @@
 {%- set htpasswd_path = "/etc/kubernetes/htpasswd" %}
 {%- set encryption_k8s_path = "/etc/kubernetes/encryption.conf" %}
 
+{# This whole block is used to "know" the Ingress external IP used by Dex.
```
I split that into a macro, why put it back here? It's a huge block of Jinja code that's not specific to API Server, hindering readability... But not critical, just questioning.

Ebaneck replied:
Yeah, I can add this split back.

The split was done in your PR that adds `Nginx certs`
Reviewer requested it to be removed since it was not used.

gdemonet commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -216,6 +216,10 @@ def _get_parts(self) -> Iterator[str]:
+    Path('salt/metalk8s/addons/dex/deployed/init.sls'),
+    Path('salt/metalk8s/addons/dex/deployed/chart.sls'),
+    Path('salt/metalk8s/addons/dex/deployed/namespace.sls'),
+
```
This is not related to the commit message, which states two things:

- "salt: deploy dex" - I would expect this commit to actually **deploy** Dex using Salt, not just prepare the buildchain
- "This commit adds the dex image to the buildchain which is required for dex deployment" - Then I don't expect anything related to Salt states in there

I think we could settle for a single commit, that handles:

- adding required images in the buildchain
- rendering `chart.sls`
- adding the few dependencies (like `tls-secret.sls`)
- adding those files in the buildchain `salt_tree.py`

The whole "Deploy Dex" feature is thus self-contained in this commit. Fixing the render script or other unrelated stuff should go in other commits (if required by this one, then go before it, if not, after).

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,98 @@
+image: '{%- endraw -%}{{ build_image_name(\"dex\", False) }}{%- raw -%}'
```
(unrelated to this line, cannot comment on the commit message)

This commit message states: "add `dex` deployment configuration", but it actually also includes the rendered chart. To me, this commit and the previous one should be squashed.

gdemonet replied:
Also, remove the note about `Secret`  not being rendered properly, as this is fixed in the same PR (just re-order commit to fix it before rendering, and squash all commits including the rendered chart).

gdemonet commented on salt/metalk8s/addons/dex/deployed/namespace.sls:
```diff
@@ -0,0 +1,10 @@
+#!jinja | metalk8s_kubernetes kubeconfig=/etc/kubernetes/admin.conf&context=kubernetes-admin@kubernetes
```
As it's directly tied to Dex and its deployment, could be squashed with the chart commit.

gdemonet commented on salt/_modules/metalk8s_network.py:
```diff
@@ -68,3 +69,13 @@ def get_cluster_dns_ip():
     return _pick_nth_service_ip(COREDNS_ADDRESS_NUMBER)
+
+
+def get_oidc_service_ip():
```
This is used in the rendered chart and chart values, so either in the same commit, or before it.

gdemonet replied:
The commit message should remove the mention about Ingress not being a valid approach, as that's what we use in the end.

gdemonet commented on salt/_modules/metalk8s_network.py:
```diff
@@ -68,3 +69,13 @@ def get_cluster_dns_ip():
+    This IP is arbitrarily selected as the seventh IP from the usable hosts
+    range.
+    '''
+    return _pick_nth_service_ip(OIDC_ADDRESS_NUMBER)
```
Missing newline here

gdemonet commented on charts/dex.yaml:
```diff
@@ -1,4 +1,4 @@
-image: '{%- endraw -%}{{ build_image_name(\"dex\", False) }}{%- raw -%}'
```
Can be squashed with the chart commit (no real value in making another commit here, the link was just for you to look at :wink:)

gdemonet commented on charts/render.py:
```diff
@@ -18,10 +18,13 @@
   `app.kubernetes.io/component` fields
 '''
 
+import re
```
(unrelated to this line, for the commit)

This commit is required before rendering the chart, so let's move it before.
Also, the commit message could be updated to:
- include a reference to the Dex deployment using charts issue (so we know that it's linked somehow)
- detail why multi-line strings are an issue in rendered charts (it's not a problem in YAML per-se, but doesn't work well when Jinja "commands" are split)

gdemonet commented on charts/dex/templates/secret.yaml:
```diff
@@ -31,7 +31,8 @@ stringData:
 {{ toYaml .connectors | indent 4 }}
     {{- end }}
-    oauth2: {{ toYaml .oauth2 | nindent 6 }}
+    oauth2:
```
This commit needs to come just after importing the chart, before rendering.

gdemonet commented on charts/dex.yaml:
```diff
@@ -32,8 +32,6 @@ ingress:
   hosts:
     - null
-  tls:
-   - secretName: dex-tls-ingress
```
No need to make this a dedicated commit, just squash with the chart rendering.

gdemonet commented on salt/metalk8s/roles/infra/init.sls:
```diff
@@ -1,2 +1,2 @@
 include:
-- metalk8s.roles.node
+  - metalk8s.roles.node
```
Unrelated to this PR, let's remove it

gdemonet commented on salt/metalk8s/salt/master/certs/init.sls:
```diff
@@ -8,8 +8,8 @@
 # * salt-api              -> generate Salt API key and certificate
 #
 include:
-  # Some server certs are required to be published in K8s API by Salt master
```
Why remove the comment? Since these aren't actual certs required by Salt master itself, it explains why they are here.

gdemonet commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -1,12 +1,14 @@
 {%- from "metalk8s/repo/macro.sls" import build_image_name with context %}
 {%- from "metalk8s/map.jinja" import networks with context %}
+{%- from "metalk8s/addons/nginx-ingress-control-plane/control-plane-ip.jinja" import ingress_control_plane with context %}
```
Line is too long to read in GH Web UI, please split at 80 chars or so

gdemonet commented on salt/metalk8s/addons/dex/deployed/tls-secret.sls:
```diff
@@ -0,0 +1,31 @@
+  ca.crt: "{{
+    salt['hashutil.base64_encodefile']('/etc/metalk8s/pki/dex/ca.crt')
+    | replace('\n', '')
+  }}"
```
Please remove this altogether, as it's not needed. (not removing it with a commit, it shouldn't appear in history at all)

gdemonet commented on charts/dex.yaml:
```diff
@@ -26,8 +26,6 @@ ingress:
   annotations:
     nginx.ingress.kubernetes.io/backend-protocol: "HTTPS"
-    nginx.ingress.kubernetes.io/proxy-ssl-verify: "on"
-    nginx.ingress.kubernetes.io/proxy-ssl-secret: "dex-ca-cert"
```
Why remove those?

Ebaneck replied:
From the above review, we remove the secret `dex-ca-cert` so it should not be needed here again I think!

gdemonet replied:
Oh right that's why I left the CA cert... My bad. Let's leave it as-is for now, open a debt ticket to set this up (and test it).

gdemonet commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -216,9 +216,15 @@ def _get_parts(self) -> Iterator[str]:
         renderer=targets.Renderer.JSON,
     ),
 
+    Path('salt/metalk8s/addons/dex/ca/init.sls'),
+    Path('salt/metalk8s/addons/dex/ca/installed.sls'),
+    Path('salt/metalk8s/addons/dex/ca/advertised.sls'),
+    Path('salt/metalk8s/addons/dex/certs/init.sls'),
+    Path('salt/metalk8s/addons/dex/certs/server.sls'),
```
Dex CA and server certificates are actually part of the Dex deployment, should be squashed with the chart render commit.

However, configuring k-a's OIDC plugin to point to Ingress, that **is** a separate commit.

Ebaneck committed:
Salt: Setup the Dex-to-Kubernetes API server integration

This commit adds the following:

Add configuration parameters to the APIserver manifest which is required
by Dex

Define a way to find the Ingress external IP required by Dex config flags

If a minion wants to reference the control-plane Ingress by its external
IP(in our case the Dex service), it needs to know the control-plane IP of
the bootstrap minion (as it is the one used by Salt master when creating the Service).

For posterity, we define a helper that should work even during
the initial boostrap.

Closes: #2010

gdemonet commented on charts/dex/templates/service.yaml:
```diff
@@ -32,6 +32,9 @@ spec:
 {{- end }}
 {{- if hasKey .Values.service "loadBalancerIP" }}
   loadBalancerIP: {{ .Values.service.loadBalancerIP }}
+{{- end }}
+{{- if hasKey .Values.service "clusterIP" }}
+  clusterIP: {{ .Values.service.clusterIP | quote }}
```
That's a manual fix on a vendored chart. It deserves a dedicated commit, explaining what is done (add support for cluster IP)

gdemonet commented on salt/metalk8s/addons/dex/ca/installed.sls:
```diff
@@ -0,0 +1,42 @@
+      - mine_function: hashutil.base64_encodefile
+      - /etc/metalk8s/pki/dex/ca.crt
+    - watch:
+      - x509: Generate dex CA certificate
```
Missing newline

gdemonet commented on salt/metalk8s/addons/dex/certs/init.sls:
```diff
@@ -0,0 +1,2 @@
+include:
+  - .server
```
Missing newline

gdemonet commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -291,6 +291,8 @@ def _get_parts(self) -> Iterator[str]:
     Path('salt/metalk8s/addons/nginx-ingress-control-plane/deployed/',
          'tls-secret.sls'),
+    Path('salt/metalk8s/addons/nginx-ingress-control-plane/',
+         'control-plane-ip.jinja'),
```
Two things:

- this commit could introduce the helper with the code that uses it, but not critical
- I don't see why it should use the `.jinja` extension, instead of `.sls` - it relies on `pillar` and `grains` (that one could setup manually, sure), so users can't expect it to work with "simple" Jinja templating

gdemonet commented on salt/metalk8s/addons/nginx-ingress-control-plane/control-plane-ip.jinja:
```diff
@@ -0,0 +1,29 @@
+      {%- set bootstrap_id = minion_id %}
+      {%- break %}
+    {%- endif %}
+  {%- endfor %}
```
This loop could be rewritten with something that we already have (forgot about it):

```jinja
{%- set bootstrap_nodes = salt.metalk8s.minions_by_role('bootstrap') %}
{%- if bootstrap_nodes %}
  {%- set bootstrap_id = bootstrap_nodes | first %}
{%- else %}
  {{ raise('Missing bootstrap node') }}
{%- endif %}
```

gdemonet reviewed (changes requested):
Mostly re-ordering of commits, and grouping things logically. The most important one is about splitting out the chart modification for support of `clusterIP`, but gathering "everything that deploys Dex" in a single commit makes sense. API Server configuration is another commit (and it's even another issue, #2010).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

* @nicolast

Ebaneck committed:
Chart: Add support for cluster IP

This commit adds support for ClusterIP in the Dex charts.

TeddyAndrieux commented on buildchain/buildchain/salt_tree.py:
```diff
@@ -216,6 +216,16 @@ def _get_parts(self) -> Iterator[str]:
     ),
 
+    Path('salt/metalk8s/addons/dex/deployed/chart.sls'),
+    Path('salt/metalk8s/addons/dex/deployed/namespace.sls'),
```
Minor: Alphabetic order you shoud move these two with the other `deployed`

TeddyAndrieux commented on salt/metalk8s/addons/nginx-ingress-control-plane/control-plane-ip.sls:
```diff
@@ -0,0 +1,28 @@
+  {%- set bootstrap_control_plane_ip = grains.metalk8s.control_plane_ip %}
+{%- else %}
+  {%- set bootstrap_control_plane_ip = salt['mine.get'](bootstrap_id,
+                                                        'control_plane_ip') %}
```
Here you have an issue you will get a dict contains control_plane_ip of each and every node so if you want to do it you need to take only the bootstrap one so

```
  {%- set bootstrap_control_plane_ip = salt['mine.get'](bootstrap_id,
                                                        'control_plane_ip')[bootstrap_id] %}
```

gdemonet commented on charts/dex/templates/service.yaml:
```diff
@@ -35,4 +35,4 @@ spec:
   selector:
     app.kubernetes.io/name: {{ include "dex.name" . }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
```
There's no reason to do this

gdemonet commented on buildchain/buildchain/constants.py:
```diff
@@ -19,6 +19,7 @@
 # URLs of the main container repositories.
 CALICO_REPOSITORY     : str = 'quay.io/calico'
 COREOS_REPOSITORY     : str = 'quay.io/coreos'
+DEX_REPOSITORY        : str = 'quay.io/dexidp'
```
Squash this with the "Deploy Dex" commit, there's no need to have a separate commit (if you reverted "Deploy Dex', then this commit still present would mean we ship unneeded container images).

gdemonet commented on charts/dex/templates/service.yaml:
```diff
@@ -32,6 +32,9 @@ spec:
 {{- end }}
 {{- if hasKey .Values.service "loadBalancerIP" }}
   loadBalancerIP: {{ .Values.service.loadBalancerIP }}
+{{- end }}
```
Commit message talks about the actual Dex cluster IP being the 7th Service IP. This is true, but not related to this commit: explanation should be added to the "Deploy Dex" commit.

gdemonet commented on charts/dex.yaml:
```diff
@@ -0,0 +1,94 @@
+      # bcrypt hash of the string "password"
+      hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
+      username: "admin"
+      userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
```
Missing newline

gdemonet commented on salt/metalk8s/addons/dex/deployed/clusterrolebinding.sls:
```diff
@@ -0,0 +1,14 @@
+roleRef:
+  kind: ClusterRole
+  name: cluster-admin
+  apiGroup: rbac.authorization.k8s.io
```
This file could have been added in another commit, or even another PR. But let's leave it, not critical.

Also, missing newline.

gdemonet reviewed (changes requested):
FYI, Dex is not an acronym, so you shouldn't write it in full-caps.

Ebaneck committed:
Certificates: Backup and restore for Dex CA-file & key

We need to backup the Dex CAfile and key in other to be able
to run a cluster restore.

This commit adds the Dex CA file and Key to the backup and restore
scripts.

Ebaneck committed:
charts: import `stable/dex` 2.19.0

```
$ helm fetch -d charts --untar stable/dex
```

Closes: #2007

Ebaneck committed:
charts: fix Dex chart to remove unneeded whitespace

That prevented a proper dumping of the Secret data

Ebaneck committed:
Salt: Add and Render Dex deployment configuration

This commit adds the following:

Adds Dex Image to the buildchain

Adds method to obtain OIDC service IP and binds this
static IP as the ClusterIP address for Dex service

Adds `metalk8-auth` namespace which holds OIDC/authentication
cluster resources(PODS)

Automatically generate Dex deployment, service account,
cluster role and clusterrolebindings

The Dex chart.sls is generated from the Helm charts using:

```
$ ./charts/render.py dex metalk8s-auth charts/dex.yaml charts/dex/ >
salt/metalk8s/addons/dex/deployed/chart.sls

Add states to deploy Dex and related server certificates

Closes: #2007
Closes: #2011

Ebaneck committed:
Salt: Bind Dex static user to a ClusterRole

The newly created Dex static user has no access to any
resources in a Metalk8s cluster.

This commit adds a clusterrolebinding which binds the Dex static
user to the Metalk8s Cluster admin Clusterrole.

Closes: #2055

gdemonet commented:
/wait

gdemonet reviewed (dismissed):
LGTM - added a wait to not merge this in 2.4.

In the meantime (waiting for 2.5 branch), could you please:
- start adding some tests (e.g. try to reach the openid-configuration and check the issuer_url)
- investigate SSL verification from the Ingress (the thing that we removed) - does it work? or is it just ignored?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

gdemonet commented:
/after_pull_request=2063

bert-e commented:
# Waiting for other pull request(s)



The current pull request is locked by the **after_pull_request** option.

In order for me to merge this pull request, run the following actions first:


:arrow_right: Merge the `OPEN` pull request:

* pull request #2063





Alternatively, delete all the **after_pull_request** comments from this pull request.





*The following options are set:* **after_pull_request**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **after_pull_request**

Ebaneck commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve, after_pull_request**

Ebaneck commented:
/status

bert-e commented:
# Status





*Status report is not available.*






*The following options are set:* **approve, after_pull_request**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast







*The following options are set:* **approve, after_pull_request**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, after_pull_request**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
